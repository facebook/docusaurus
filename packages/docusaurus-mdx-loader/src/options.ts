/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {MDXOptions, SimpleProcessors} from './processor';
import type {MarkdownConfig} from '@docusaurus/types';
import type {ResolveMarkdownLink} from './remark/resolveMarkdownLinks';
import type {PromiseWithResolvers} from './utils';

export type Options = Partial<MDXOptions> & {
  markdownConfig: MarkdownConfig;
  staticDirs: string[];
  siteDir: string;
  isMDXPartial?: (filePath: string) => boolean;
  isMDXPartialFrontMatterWarningDisabled?: boolean;
  removeContentTitle?: boolean;
  metadataPath?: (filePath: string) => string;
  createAssets?: (metadata: {
    filePath: string;
    frontMatter: {[key: string]: unknown};
  }) => {[key: string]: unknown};
  resolveMarkdownLink?: ResolveMarkdownLink;

  // Will usually be created by "createMDXLoaderItem"
  processors?: SimpleProcessors;
  crossCompilerCache?: Map<string, CrossCompilerCacheEntry>; // MDX => Promise<JSX> cache
};

type CrossCompilerCacheEntry = PromiseWithResolvers<string>;
