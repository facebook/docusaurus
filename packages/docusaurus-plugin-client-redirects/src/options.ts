/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Joi, PathnameSchema} from '@docusaurus/utils-validation';
import type {OptionValidationContext} from '@docusaurus/types';

export type RedirectOption = {
  /** Pathname of an existing Docusaurus page */
  to: string;
  /** Pathname of the new page(s) we should create */
  from: string | string[];
};

export type PluginOptions = {
  /** Plugin ID. */
  id: string;
  /** The extensions to be removed from the route after redirecting. */
  fromExtensions: string[];
  /** The extensions to be appended to the route after redirecting. */
  toExtensions: string[];
  /** The list of redirect rules, each one with multiple `from`s â†’ one `to`. */
  redirects: RedirectOption[];
  /**
   * A callback to create a redirect rule. Docusaurus query this callback
   * against every path it has created, and use its return value to output more
   * paths.
   * @returns All the paths from which we should redirect to `path`
   */
  createRedirects?: (
    /** An existing Docusaurus route path */
    path: string,
  ) => string[] | string | null | undefined;
};

export type Options = Partial<PluginOptions>;

export const DEFAULT_OPTIONS: Partial<PluginOptions> = {
  fromExtensions: [],
  toExtensions: [],
  redirects: [],
};

const RedirectPluginOptionValidation = Joi.object<RedirectOption>({
  from: Joi.alternatives().try(
    PathnameSchema.required(),
    Joi.array().items(PathnameSchema.required()),
  ),
  to: Joi.string().required(),
});

const isString = Joi.string().required().not(null);

const UserOptionsSchema = Joi.object<PluginOptions>({
  fromExtensions: Joi.array()
    .items(isString)
    .default(DEFAULT_OPTIONS.fromExtensions),
  toExtensions: Joi.array()
    .items(isString)
    .default(DEFAULT_OPTIONS.toExtensions),
  redirects: Joi.array()
    .items(RedirectPluginOptionValidation)
    .default(DEFAULT_OPTIONS.redirects),
  createRedirects: Joi.function().maxArity(1),
}).default(DEFAULT_OPTIONS);

export function validateOptions({
  validate,
  options: userOptions,
}: OptionValidationContext<Options | undefined, PluginOptions>): PluginOptions {
  return validate(UserOptionsSchema, userOptions);
}
