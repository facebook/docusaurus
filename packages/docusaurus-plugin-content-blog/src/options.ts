/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {
  Joi,
  RemarkPluginsSchema,
  RehypePluginsSchema,
  RecmaPluginsSchema,
  AdmonitionsSchema,
  RouteBasePathSchema,
  URISchema,
} from '@docusaurus/utils-validation';
import {GlobExcludeDefault} from '@docusaurus/utils';
import type {
  PluginOptions,
  Options,
  FeedType,
  FeedXSLTOptions,
} from '@docusaurus/plugin-content-blog';
import type {OptionValidationContext} from '@docusaurus/types';

export const DEFAULT_OPTIONS: PluginOptions = {
  feedOptions: {
    type: ['rss', 'atom'],
    copyright: '',
    limit: 20,
    xslt: {
      rss: null,
      atom: null,
    },
  },
  beforeDefaultRehypePlugins: [],
  beforeDefaultRemarkPlugins: [],
  admonitions: true,
  truncateMarker: /<!--\s*truncate\s*-->|\{\/\*\s*truncate\s*\*\/\}/,
  rehypePlugins: [],
  remarkPlugins: [],
  recmaPlugins: [],
  showReadingTime: true,
  blogTagsPostsComponent: '@theme/BlogTagsPostsPage',
  blogTagsListComponent: '@theme/BlogTagsListPage',
  blogAuthorsPostsComponent: '@theme/Blog/Pages/BlogAuthorsPostsPage',
  blogAuthorsListComponent: '@theme/Blog/Pages/BlogAuthorsListPage',
  blogPostComponent: '@theme/BlogPostPage',
  blogListComponent: '@theme/BlogListPage',
  blogArchiveComponent: '@theme/BlogArchivePage',
  blogDescription: 'Blog',
  blogTitle: 'Blog',
  blogSidebarCount: 5,
  blogSidebarTitle: 'Recent posts',
  postsPerPage: 10,
  include: ['**/*.{md,mdx}'],
  exclude: GlobExcludeDefault,
  routeBasePath: 'blog',
  tagsBasePath: 'tags',
  archiveBasePath: 'archive',
  pageBasePath: 'page',
  path: 'blog',
  editLocalizedFiles: false,
  authorsMapPath: 'authors.yml',
  readingTime: ({content, defaultReadingTime}) => defaultReadingTime({content}),
  sortPosts: 'descending',
  showLastUpdateTime: false,
  showLastUpdateAuthor: false,
  processBlogPosts: async () => undefined,
  tags: undefined,
  authorsBasePath: 'authors',
  onInlineTags: 'warn',
  onInlineAuthors: 'warn',
  onUntruncatedBlogPosts: 'warn',
};

export const XSLTBuiltInPaths = {
  rss: path.resolve(__dirname, '..', 'assets', 'rss.xsl'),
  atom: path.resolve(__dirname, '..', 'assets', 'atom.xsl'),
};

function normalizeXsltOption(
  option: string | null | boolean,
  type: 'rss' | 'atom',
): string | null {
  if (typeof option === 'string') {
    return option;
  }
  if (option === true) {
    return XSLTBuiltInPaths[type];
  }
  return null;
}

function createXSLTFilePathSchema(type: 'atom' | 'rss') {
  return Joi.alternatives()
    .try(
      Joi.string().required(),
      Joi.boolean()
        .allow(null, () => undefined)
        .custom((val) => normalizeXsltOption(val, type)),
    )
    .optional()
    .default(null);
}

const FeedXSLTOptionsSchema = Joi.alternatives()
  .try(
    Joi.object<FeedXSLTOptions>({
      rss: createXSLTFilePathSchema('rss'),
      atom: createXSLTFilePathSchema('atom'),
    }).required(),
    Joi.boolean()
      .allow(null, () => undefined)
      .custom((val) => ({
        rss: normalizeXsltOption(val, 'rss'),
        atom: normalizeXsltOption(val, 'atom'),
      })),
  )
  .optional()
  .custom((val) => {
    if (val === null) {
      return {
        rss: null,
        atom: null,
      };
    }
    return val;
  })
  .default(DEFAULT_OPTIONS.feedOptions.xslt);

const FeedOptionsSchema = Joi.object({
  type: Joi.alternatives()
    .try(
      Joi.array().items(Joi.string().equal('rss', 'atom', 'json')),
      Joi.alternatives().conditional(
        Joi.string().equal('all', 'rss', 'atom', 'json'),
        {
          then: Joi.custom((val: FeedType | 'all') =>
            val === 'all' ? ['rss', 'atom', 'json'] : [val],
          ),
        },
      ),
    )
    .allow(null)
    .default(DEFAULT_OPTIONS.feedOptions.type),
  xslt: FeedXSLTOptionsSchema,
  title: Joi.string().allow(''),
  description: Joi.string().allow(''),
  // Only add default value when user actually wants a feed (type is not null)
  copyright: Joi.when('type', {
    is: Joi.any().valid(null),
    then: Joi.string().optional(),
    otherwise: Joi.string()
      .allow('')
      .default(DEFAULT_OPTIONS.feedOptions.copyright),
  }),
  language: Joi.string(),
  createFeedItems: Joi.function(),
  limit: Joi.alternatives()
    .try(Joi.number(), Joi.valid(null), Joi.valid(false))
    .default(DEFAULT_OPTIONS.feedOptions.limit),
}).default(DEFAULT_OPTIONS.feedOptions);

const PluginOptionSchema = Joi.object<PluginOptions>({
  path: Joi.string().default(DEFAULT_OPTIONS.path),
  archiveBasePath: Joi.string()
    .default(DEFAULT_OPTIONS.archiveBasePath)
    .allow(null),
  routeBasePath: RouteBasePathSchema.default(DEFAULT_OPTIONS.routeBasePath),
  tagsBasePath: Joi.string().default(DEFAULT_OPTIONS.tagsBasePath),
  pageBasePath: Joi.string().default(DEFAULT_OPTIONS.pageBasePath),
  include: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.include),
  exclude: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.exclude),
  postsPerPage: Joi.alternatives()
    .try(Joi.equal('ALL').required(), Joi.number().integer().min(1).required())
    .default(DEFAULT_OPTIONS.postsPerPage),
  blogListComponent: Joi.string().default(DEFAULT_OPTIONS.blogListComponent),
  blogPostComponent: Joi.string().default(DEFAULT_OPTIONS.blogPostComponent),
  blogTagsListComponent: Joi.string().default(
    DEFAULT_OPTIONS.blogTagsListComponent,
  ),
  blogTagsPostsComponent: Joi.string().default(
    DEFAULT_OPTIONS.blogTagsPostsComponent,
  ),
  blogAuthorsPostsComponent: Joi.string().default(
    DEFAULT_OPTIONS.blogAuthorsPostsComponent,
  ),
  blogAuthorsListComponent: Joi.string().default(
    DEFAULT_OPTIONS.blogAuthorsListComponent,
  ),
  blogArchiveComponent: Joi.string().default(
    DEFAULT_OPTIONS.blogArchiveComponent,
  ),
  blogTitle: Joi.string().allow('').default(DEFAULT_OPTIONS.blogTitle),
  blogDescription: Joi.string()
    .allow('')
    .default(DEFAULT_OPTIONS.blogDescription),
  blogSidebarCount: Joi.alternatives()
    .try(Joi.equal('ALL').required(), Joi.number().integer().min(0).required())
    .default(DEFAULT_OPTIONS.blogSidebarCount),
  blogSidebarTitle: Joi.string().default(DEFAULT_OPTIONS.blogSidebarTitle),
  showReadingTime: Joi.bool().default(DEFAULT_OPTIONS.showReadingTime),
  remarkPlugins: RemarkPluginsSchema.default(DEFAULT_OPTIONS.remarkPlugins),
  rehypePlugins: RehypePluginsSchema.default(DEFAULT_OPTIONS.rehypePlugins),
  recmaPlugins: RecmaPluginsSchema.default(DEFAULT_OPTIONS.recmaPlugins),
  admonitions: AdmonitionsSchema.default(DEFAULT_OPTIONS.admonitions),
  editUrl: Joi.alternatives().try(URISchema, Joi.function()),
  editLocalizedFiles: Joi.boolean().default(DEFAULT_OPTIONS.editLocalizedFiles),
  truncateMarker: Joi.object().default(DEFAULT_OPTIONS.truncateMarker),
  beforeDefaultRemarkPlugins: RemarkPluginsSchema.default(
    DEFAULT_OPTIONS.beforeDefaultRemarkPlugins,
  ),
  beforeDefaultRehypePlugins: RehypePluginsSchema.default(
    DEFAULT_OPTIONS.beforeDefaultRehypePlugins,
  ),
  feedOptions: FeedOptionsSchema,
  authorsMapPath: Joi.string().default(DEFAULT_OPTIONS.authorsMapPath),
  readingTime: Joi.function().default(() => DEFAULT_OPTIONS.readingTime),
  sortPosts: Joi.string()
    .valid('descending', 'ascending')
    .default(DEFAULT_OPTIONS.sortPosts),
  showLastUpdateTime: Joi.bool().default(DEFAULT_OPTIONS.showLastUpdateTime),
  showLastUpdateAuthor: Joi.bool().default(
    DEFAULT_OPTIONS.showLastUpdateAuthor,
  ),
  processBlogPosts: Joi.function()
    .optional()
    .default(() => DEFAULT_OPTIONS.processBlogPosts),
  onInlineTags: Joi.string()
    .equal('ignore', 'log', 'warn', 'throw')
    .default(DEFAULT_OPTIONS.onInlineTags),
  tags: Joi.string()
    .disallow('')
    .allow(null, false)
    .default(() => DEFAULT_OPTIONS.tags),
  authorsBasePath: Joi.string()
    .default(DEFAULT_OPTIONS.authorsBasePath)
    .disallow(''),
  onInlineAuthors: Joi.string()
    .equal('ignore', 'log', 'warn', 'throw')
    .default(DEFAULT_OPTIONS.onInlineAuthors),
  onUntruncatedBlogPosts: Joi.string()
    .equal('ignore', 'log', 'warn', 'throw')
    .default(DEFAULT_OPTIONS.onUntruncatedBlogPosts),
}).default(DEFAULT_OPTIONS);

export function validateOptions({
  validate,
  options,
}: OptionValidationContext<Options | undefined, PluginOptions>): PluginOptions {
  const validatedOptions = validate(PluginOptionSchema, options);
  return validatedOptions;
}
