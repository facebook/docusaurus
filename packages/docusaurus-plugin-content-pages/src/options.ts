/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  Joi,
  RemarkPluginsSchema,
  RehypePluginsSchema,
  RecmaPluginsSchema,
  AdmonitionsSchema,
  RouteBasePathSchema,
  URISchema,
} from '@docusaurus/utils-validation';
import {GlobExcludeDefault} from '@docusaurus/utils';
import type {OptionValidationContext} from '@docusaurus/types';
import type {PluginOptions, Options} from '@docusaurus/plugin-content-pages';

export const DEFAULT_OPTIONS: PluginOptions = {
  path: 'src/pages', // Path to data on filesystem, relative to site dir.
  routeBasePath: '/', // URL Route.
  include: ['**/*.{js,jsx,ts,tsx,md,mdx}'], // Extensions to include.
  exclude: GlobExcludeDefault,
  mdxPageComponent: '@theme/MDXPage',
  remarkPlugins: [],
  rehypePlugins: [],
  recmaPlugins: [],
  beforeDefaultRehypePlugins: [],
  beforeDefaultRemarkPlugins: [],
  admonitions: true,
  showLastUpdateTime: false,
  showLastUpdateAuthor: false,
  editLocalizedFiles: false,
};

const PluginOptionSchema = Joi.object<PluginOptions>({
  path: Joi.string().default(DEFAULT_OPTIONS.path),
  routeBasePath: RouteBasePathSchema.default(DEFAULT_OPTIONS.routeBasePath),
  include: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.include),
  exclude: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.exclude),
  mdxPageComponent: Joi.string().default(DEFAULT_OPTIONS.mdxPageComponent),
  remarkPlugins: RemarkPluginsSchema.default(DEFAULT_OPTIONS.remarkPlugins),
  rehypePlugins: RehypePluginsSchema.default(DEFAULT_OPTIONS.rehypePlugins),
  recmaPlugins: RecmaPluginsSchema.default(DEFAULT_OPTIONS.recmaPlugins),
  beforeDefaultRehypePlugins: RehypePluginsSchema.default(
    DEFAULT_OPTIONS.beforeDefaultRehypePlugins,
  ),
  beforeDefaultRemarkPlugins: RemarkPluginsSchema.default(
    DEFAULT_OPTIONS.beforeDefaultRemarkPlugins,
  ),
  admonitions: AdmonitionsSchema.default(DEFAULT_OPTIONS.admonitions),
  showLastUpdateTime: Joi.bool().default(DEFAULT_OPTIONS.showLastUpdateTime),
  showLastUpdateAuthor: Joi.bool().default(
    DEFAULT_OPTIONS.showLastUpdateAuthor,
  ),
  editUrl: Joi.alternatives().try(URISchema, Joi.function()),
  editLocalizedFiles: Joi.boolean().default(DEFAULT_OPTIONS.editLocalizedFiles),
});

export function validateOptions({
  validate,
  options,
}: OptionValidationContext<Options, PluginOptions>): PluginOptions {
  const validatedOptions = validate(PluginOptionSchema, options);
  return validatedOptions;
}
