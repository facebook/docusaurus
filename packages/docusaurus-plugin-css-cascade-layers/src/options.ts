/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import {Joi} from '@docusaurus/utils-validation';
import {posixPath} from '@docusaurus/utils';
import {isValidLayerName} from './layers';
import type {OptionValidationContext} from '@docusaurus/types';

export type PluginOptions = {
  id: string; // plugin id
  layers: Record<string, (filePath: string) => boolean>;
};

export type Options = {
  layers?: PluginOptions['layers'];
};

// Not ideal to compute layers using "filePath.includes()"
// But this is mostly temporary until we add first-class layers everywhere
function layerFor(...params: string[]) {
  return (filePath: string) => {
    const posixFilePath = posixPath(filePath);
    return params.some((p) => posixFilePath.includes(p));
  };
}

// Object order matters, it defines the layer order
export const DEFAULT_LAYERS: PluginOptions['layers'] = {
  'docusaurus.infima': layerFor('node_modules/infima/dist'),
  'docusaurus.theme-common': layerFor(
    'packages/docusaurus-theme-common/lib',
    'node_modules/@docusaurus/theme-common/lib',
  ),
  'docusaurus.theme-classic': layerFor(
    'packages/docusaurus-theme-classic/lib',
    'node_modules/@docusaurus/theme-classic/lib',
  ),
  'docusaurus.core': layerFor(
    'packages/docusaurus/lib',
    'node_modules/@docusaurus/core/lib',
  ),
  'docusaurus.plugin-debug': layerFor(
    'packages/docusaurus-plugin-debug/lib',
    'node_modules/@docusaurus/plugin-debug/lib',
  ),
  'docusaurus.theme-mermaid': layerFor(
    'packages/docusaurus-theme-mermaid/lib',
    'node_modules/@docusaurus/theme-mermaid/lib',
  ),
  'docusaurus.theme-live-codeblock': layerFor(
    'packages/docusaurus-theme-live-codeblock/lib',
    'node_modules/@docusaurus/theme-live-codeblock/lib',
  ),
  'docusaurus.theme-search-algolia.docsearch': layerFor(
    'node_modules/@docsearch/css/dist',
  ),
  'docusaurus.theme-search-algolia': layerFor(
    'packages/docusaurus-theme-search-algolia/lib',
    'node_modules/@docusaurus/theme-search-algolia/lib',
  ),
  // docusaurus.website layer ? (declare it, even if empty?)
};

export const DEFAULT_OPTIONS: Partial<PluginOptions> = {
  id: 'default',
  layers: DEFAULT_LAYERS,
};

const pluginOptionsSchema = Joi.object<PluginOptions>({
  layers: Joi.object()
    .pattern(
      Joi.custom((val, helpers) => {
        if (!isValidLayerName(val)) {
          return helpers.error('any.invalid');
        }
        return val;
      }),
      Joi.function().arity(1).required(),
    )
    .default(DEFAULT_LAYERS),
});

export function validateOptions({
  validate,
  options,
}: OptionValidationContext<Options, PluginOptions>): PluginOptions {
  return validate(pluginOptionsSchema, options);
}
