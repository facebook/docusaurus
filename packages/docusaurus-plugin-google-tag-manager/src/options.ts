/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export type PluginOptions = {
  containerId: string;
};

export type Options = Partial<PluginOptions>;
