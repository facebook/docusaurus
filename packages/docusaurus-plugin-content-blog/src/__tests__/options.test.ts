/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {normalizePluginOptions} from '@docusaurus/utils-validation';
import {validateOptions, DEFAULT_OPTIONS, XSLTBuiltInPaths} from '../options';
import type {
  Options,
  PluginOptions,
  UserFeedOptions,
} from '@docusaurus/plugin-content-blog';
import type {Validate} from '@docusaurus/types';

function testValidate(options?: Options) {
  return validateOptions({
    validate: normalizePluginOptions as Validate<
      Options | undefined,
      PluginOptions
    >,
    options,
  });
}

// The type of remark/rehype plugins can be either function, object or array
const markdownPluginsFunctionStub = () => {};
const markdownPluginsObjectStub = {};

const defaultOptions = {...DEFAULT_OPTIONS, id: 'default'};

describe('validateOptions', () => {
  it('returns default options for undefined user options', () => {
    expect(testValidate(undefined)).toEqual(defaultOptions);
  });

  it('returns default options for empty user options', () => {
    expect(testValidate({})).toEqual(defaultOptions);
  });

  it('accepts correctly defined user options', () => {
    const userOptions: Options = {
      ...defaultOptions,
      feedOptions: {
        type: 'rss' as const,
        title: 'myTitle',
      },
      path: 'not_blog',
      routeBasePath: '/myBlog',
      postsPerPage: 5,
      include: ['api/*', 'docs/*'],
      tags: 'customTags.yml',
      onInlineTags: 'warn',
    };
    expect(testValidate(userOptions)).toEqual({
      ...userOptions,
      feedOptions: {
        type: ['rss'],
        title: 'myTitle',
        copyright: '',
        limit: 20,
        xslt: {rss: null, atom: null},
      },
    });
  });

  it('accepts valid user options', () => {
    const userOptions: Options = {
      ...defaultOptions,
      routeBasePath: '/myBlog',
      beforeDefaultRemarkPlugins: [],
      beforeDefaultRehypePlugins: [markdownPluginsFunctionStub],
      remarkPlugins: [[markdownPluginsFunctionStub, {option1: '42'}]],
      rehypePlugins: [
        markdownPluginsObjectStub,
        [markdownPluginsFunctionStub, {option1: '42'}],
      ],
      recmaPlugins: [markdownPluginsFunctionStub],
    };
    expect(testValidate(userOptions)).toEqual(userOptions);
  });

  it('throws Error in case of invalid options', () => {
    expect(() =>
      testValidate({
        path: 'not_blog',
        postsPerPage: -1,
        include: ['api/*', 'docs/*'],
        routeBasePath: 'not_blog',
      }),
    ).toThrowErrorMatchingSnapshot();
  });

  describe('feed', () => {
    it('throws Error in case of invalid feed type', () => {
      expect(() =>
        testValidate({
          feedOptions: {
            // @ts-expect-error: test
            type: 'none',
          },
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    it('converts all feed type to array with other feed type', () => {
      expect(
        testValidate({
          feedOptions: {type: 'all'},
        }),
      ).toEqual({
        ...defaultOptions,
        feedOptions: {
          type: ['rss', 'atom', 'json'],
          copyright: '',
          limit: 20,
          xslt: {rss: null, atom: null},
        },
      });
    });

    it('accepts null feed type and return same', () => {
      expect(
        testValidate({
          feedOptions: {type: null},
        }),
      ).toEqual({
        ...defaultOptions,
        feedOptions: {
          type: null,
          limit: 20,
          xslt: {rss: null, atom: null},
        },
      });
    });

    it('contains array with rss + atom for missing feed type', () => {
      expect(
        testValidate({
          feedOptions: {},
        }),
      ).toEqual(defaultOptions);
    });

    it('has array with rss + atom, title for missing feed type', () => {
      expect(
        testValidate({
          feedOptions: {title: 'title'},
        }),
      ).toEqual({
        ...defaultOptions,
        feedOptions: {
          type: ['rss', 'atom'],
          title: 'title',
          copyright: '',
          limit: 20,
          xslt: {rss: null, atom: null},
        },
      });
    });

    describe('feed xslt', () => {
      function testXSLT(xslt: UserFeedOptions['xslt']) {
        return testValidate({feedOptions: {xslt}}).feedOptions.xslt;
      }

      it('accepts xslt: true', () => {
        expect(testXSLT(true)).toEqual({
          rss: XSLTBuiltInPaths.rss,
          atom: XSLTBuiltInPaths.atom,
        });
      });

      it('accepts xslt: false', () => {
        expect(testXSLT(false)).toEqual({
          rss: null,
          atom: null,
        });
      });

      it('accepts xslt: null', () => {
        expect(testXSLT(null)).toEqual({
          rss: null,
          atom: null,
        });
      });

      it('accepts xslt: undefined', () => {
        expect(testXSLT(undefined)).toEqual({
          rss: null,
          atom: null,
        });
      });

      it('accepts xslt: {rss: true}', () => {
        expect(testXSLT({rss: true})).toEqual({
          rss: XSLTBuiltInPaths.rss,
          atom: null,
        });
      });

      it('accepts xslt: {atom: true}', () => {
        expect(testXSLT({atom: true})).toEqual({
          rss: null,
          atom: XSLTBuiltInPaths.atom,
        });
      });

      it('accepts xslt: {rss: true, atom: true}', () => {
        expect(testXSLT({rss: true, atom: true})).toEqual({
          rss: XSLTBuiltInPaths.rss,
          atom: XSLTBuiltInPaths.atom,
        });
      });

      it('accepts xslt: {rss: "custom-path"}', () => {
        expect(testXSLT({rss: 'custom-path'})).toEqual({
          rss: 'custom-path',
          atom: null,
        });
      });

      it('accepts xslt: {rss: true, atom: "custom-path"}', () => {
        expect(testXSLT({rss: true, atom: 'custom-path'})).toEqual({
          rss: XSLTBuiltInPaths.rss,
          atom: 'custom-path',
        });
      });

      it('accepts xslt: {rss: null, atom: true}', () => {
        expect(testXSLT({rss: null, atom: true})).toEqual({
          rss: null,
          atom: XSLTBuiltInPaths.atom,
        });
      });

      it('accepts xslt: {rss: false, atom: null}', () => {
        expect(testXSLT({rss: false, atom: null})).toEqual({
          rss: null,
          atom: null,
        });
      });

      it('rejects xslt: 42', () => {
        // @ts-expect-error: bad type
        expect(() => testXSLT(42)).toThrowErrorMatchingInlineSnapshot(
          `""feedOptions.xslt" must be one of [object, boolean]"`,
        );
      });
      it('rejects xslt: []', () => {
        // @ts-expect-error: bad type
        expect(() => testXSLT([])).toThrowErrorMatchingInlineSnapshot(
          `""feedOptions.xslt" must be one of [object, boolean]"`,
        );
      });

      it('rejects xslt: {rss: 42}', () => {
        // @ts-expect-error: bad type
        expect(() => testXSLT({rss: 42})).toThrowErrorMatchingInlineSnapshot(
          `""feedOptions.xslt.rss" must be one of [string, boolean]"`,
        );
      });

      it('rejects xslt: {rss: []}', () => {
        // @ts-expect-error: bad type
        expect(() => testXSLT({rss: 42})).toThrowErrorMatchingInlineSnapshot(
          `""feedOptions.xslt.rss" must be one of [string, boolean]"`,
        );
      });
    });
  });

  it('accepts 0 sidebar count', () => {
    const userOptions = {blogSidebarCount: 0};
    expect(testValidate(userOptions)).toEqual({
      ...defaultOptions,
      ...userOptions,
    });
  });

  it('accepts "ALL" sidebar count', () => {
    const userOptions = {blogSidebarCount: 'ALL' as const};
    expect(testValidate(userOptions)).toEqual({
      ...defaultOptions,
      ...userOptions,
    });
  });

  it('rejects "abcdef" sidebar count', () => {
    const userOptions = {blogSidebarCount: 'abcdef'};
    // @ts-expect-error: test
    expect(() => testValidate(userOptions)).toThrowErrorMatchingInlineSnapshot(
      `""blogSidebarCount" must be one of [ALL, number]"`,
    );
  });

  it('accepts "all posts" sidebar title', () => {
    const userOptions = {blogSidebarTitle: 'all posts'};
    expect(testValidate(userOptions)).toEqual({
      ...defaultOptions,
      ...userOptions,
    });
  });

  it('rejects 42 sidebar title', () => {
    const userOptions = {blogSidebarTitle: 42};
    // @ts-expect-error: test
    expect(() => testValidate(userOptions)).toThrowErrorMatchingInlineSnapshot(
      `""blogSidebarTitle" must be a string"`,
    );
  });

  describe('tags', () => {
    it('accepts tags - undefined', () => {
      expect(testValidate({tags: undefined}).tags).toBeUndefined();
    });

    it('accepts tags - null', () => {
      expect(testValidate({tags: null}).tags).toBeNull();
    });

    it('accepts tags - false', () => {
      expect(testValidate({tags: false}).tags).toBeFalsy();
    });

    it('accepts tags - customTags.yml', () => {
      expect(testValidate({tags: 'customTags.yml'}).tags).toBe(
        'customTags.yml',
      );
    });

    it('rejects tags - 42', () => {
      // @ts-expect-error: test
      expect(() => testValidate({tags: 42})).toThrowErrorMatchingInlineSnapshot(
        `""tags" must be a string"`,
      );
    });
  });

  describe('onInlineTags', () => {
    it('accepts onInlineTags - undefined', () => {
      expect(testValidate({onInlineTags: undefined}).onInlineTags).toBe('warn');
    });

    it('accepts onInlineTags - "throw"', () => {
      expect(testValidate({onInlineTags: 'throw'}).onInlineTags).toBe('throw');
    });

    it('rejects onInlineTags - "trace"', () => {
      expect(() =>
        // @ts-expect-error: test
        testValidate({onInlineTags: 'trace'}),
      ).toThrowErrorMatchingInlineSnapshot(
        `""onInlineTags" must be one of [ignore, log, warn, throw]"`,
      );
    });

    it('rejects onInlineTags - null', () => {
      expect(() =>
        // @ts-expect-error: test
        testValidate({onInlineTags: 42}),
      ).toThrowErrorMatchingInlineSnapshot(
        `""onInlineTags" must be one of [ignore, log, warn, throw]"`,
      );
    });

    it('rejects onInlineTags - 42', () => {
      expect(() =>
        // @ts-expect-error: test
        testValidate({onInlineTags: 42}),
      ).toThrowErrorMatchingInlineSnapshot(
        `""onInlineTags" must be one of [ignore, log, warn, throw]"`,
      );
    });
  });

  describe('onUntruncatedBlogPosts', () => {
    it('accepts onUntruncatedBlogPosts - undefined', () => {
      expect(
        testValidate({onUntruncatedBlogPosts: undefined})
          .onUntruncatedBlogPosts,
      ).toBe('warn');
    });

    it('accepts onUntruncatedBlogPosts - "throw"', () => {
      expect(
        testValidate({onUntruncatedBlogPosts: 'throw'}).onUntruncatedBlogPosts,
      ).toBe('throw');
    });

    it('rejects onUntruncatedBlogPosts - "trace"', () => {
      expect(() =>
        // @ts-expect-error: test
        testValidate({onUntruncatedBlogPosts: 'trace'}),
      ).toThrowErrorMatchingInlineSnapshot(
        `""onUntruncatedBlogPosts" must be one of [ignore, log, warn, throw]"`,
      );
    });

    it('rejects onUntruncatedBlogPosts - null', () => {
      expect(() =>
        // @ts-expect-error: test
        testValidate({onUntruncatedBlogPosts: 42}),
      ).toThrowErrorMatchingInlineSnapshot(
        `""onUntruncatedBlogPosts" must be one of [ignore, log, warn, throw]"`,
      );
    });

    it('rejects onUntruncatedBlogPosts - 42', () => {
      expect(() =>
        // @ts-expect-error: test
        testValidate({onUntruncatedBlogPosts: 42}),
      ).toThrowErrorMatchingInlineSnapshot(
        `""onUntruncatedBlogPosts" must be one of [ignore, log, warn, throw]"`,
      );
    });
  });
});
