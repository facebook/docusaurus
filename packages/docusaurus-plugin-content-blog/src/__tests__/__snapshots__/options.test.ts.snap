// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateOptions feed throws Error in case of invalid feed type 1`] = `""feedOptions.type" does not match any of the allowed types"`;

exports[`validateOptions throws Error in case of invalid options 1`] = `""postsPerPage" must be greater than or equal to 1"`;
